mlflow/xgboost.py

```diff
@@ -365,12 +369,39 @@ def record_eval_results(eval_results, metrics_logger):
             Create a callback function that records evaluation results.
             """
 
-            @exception_safe_function
-            def callback(env):
-                metrics_logger.record_metrics(dict(env.evaluation_result_list), env.iteration)
-                eval_results.append(dict(env.evaluation_result_list))
+            if LooseVersion(xgboost.__version__) >= LooseVersion("1.3.0"):
+                # In xgboost >= 1.3.0, user-defined callbacks should inherit
+                # `xgboost.callback.TrainingCallback`:
+                # https://xgboost.readthedocs.io/en/latest/python/callbacks.html#defining-your-own-callback  # noqa
 
-            return callback
+                class Callback(
+                    xgboost.callback.TrainingCallback, metaclass=ExceptionSafeAbstractClass,
+                ):
+                    def after_iteration(self, model, epoch, evals_log):
+                        """
+                        Run after each iteration. Return True when training should stop.
+                        """
+                        evaluation_result_dict = {}
+                        for data_name, metric_dict in evals_log.items():
+                            for metric_name, metric_values in metric_dict.items():
+                                key = "{}-{}".format(data_name, metric_name)
+                                evaluation_result_dict[key] = metric_values[-1]
```

Comment:
`metric_values` is a list of metrics on each iteration. The last element corresonds the metric on the current iteration.

Reply:
it looksl like:

````python
on iter 1: {"train", {"metric": [0.1], ...}, "valid": ...}
on iter 2: {"train", {"metric": [0.1, 0.2], ...}, "valid": ...}
on iter 3: {"train", {"metric": [0.1, 0.2, 0.3], ...}, "valid": ...}

Reply:
Awesome! Can we add some inline comments explaining the logic?

Reply:
A bit confused about the data structure in your example. Is this a set? should `valid` be nested under the `metric_dict`?

Reply:
Fixed the syntax. It's not a set but a dict.

Reply:
Added some comments to clarify the logic here